### PR TITLE
[SYCL] FE support for optional device features.

### PR DESCRIPTION
**Pull request by elizabethandrews**
This patch implements 2 new attributes [[sycl::device_has()]]
and [[__sycl_detail__::__uses_aspects__()]]. Both attributes
accept 0 or more device aspects (cl::sycl::aspect) as arguments.

[[sycl::device_has()]] can be applied only to functions. If a
function is decorated with this attribute, !intel_declared_aspects
metadata, listing the aspects specified in attribute argument, is
added to the function metadata.

[[__sycl_detail__::__uses_aspects__()]] is an undocumented internal
attribute which can be applied to records (type declarations) and
function. If a function is decorated with this attribute,
!intel_used_aspects metadata, listing the aspects specified in
attribute argument, is added to the function metadata. If a record
is decorated with this attribute, a named metadata node
!intel_types_that_use_aspects is generated. The value of the this
metadata is a list of unnamed metadata nodes, each of which
describes one type that is decorated with this attribute. The value
of each unnamed metadata node starts with a string giving the name
of the type which is followed by a list of i32 constants, where each
constant is a value from cl::sycl::aspect enum class, telling the
numerical value of an aspect from the type's
[[sycl_detail::uses_aspects()]] attribute.

E.g.
!intel_types_that_use_aspects = !{!0, !1, !2}
!0 = !{!"class.cl::sycl::detail::half_impl::half", i32 8}
!1 = !{!"class.cl::sycl::amx_type", i32 9}
!2 = !{!"class.cl::sycl::other_type", i32 8, i32 9}

In the example above, the type cl::sycl::detail::half_impl::half uses
an aspect whose numerical value is 8 and the type cl::sycl::other_type
uses two aspects 8 and 9.

Spec:
github.com/intel/llvm/blob/sycl/sycl/doc/OptionalDeviceFeatures.md

Signed-off-by: Elizabeth Andrews <elizabeth.andrews@intel.com>

### CODE REVIEW

File: clang/include/clang/Basic/AttrDocs.td
Comment (smanna12) [inline clang/include/clang/Basic/AttrDocs.td:450]: This seems a typo. Did you meant to use [[sycl::device_has()]] int func() {}?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (smanna12) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:11528]: Can we update diagnostic message like -
def err_sycl_invalid_aspect_argument : Error<
  "%0 attribute argument is invalid; must be device aspect of type cl::sycl::aspect">;

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (smanna12) [inline clang/lib/CodeGen/CodeGenFunction.cpp:1058]:
```suggestion
   if (const auto *A = D->getAttr<SYCLDeviceHasAttr>()) {
```

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (smanna12) [inline clang/lib/CodeGen/CodeGenFunction.cpp:1068]:
```suggestion
    if (const auto *A = D->getAttr<SYCLUsesAspectsAttr>()) {
```

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [inline clang/lib/Sema/SemaDeclAttr.cpp:9401]:
```suggestion
  if (const auto *ExistingAttr = D->getAttr<SYCLUsesAspectsAttr>()) {
```

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [inline clang/lib/Sema/SemaDeclAttr.cpp:9368]:
```suggestion
  if (const auto *ExistingAttr = D->getAttr<SYCLDeviceHasAttr>()) {
```

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclAttr.cpp:9344]: I may have seen code this function contains somewhere around SemaSYCL. Should we avoid the duplication?

File: clang/lib/CodeGen/CodeGenModule.h
Comment (Fznamznon) [inline clang/lib/CodeGen/CodeGenModule.h:1053]: NIT
```suggestion
  void addTypeWithAspects(StringRef TypeName, const RecordDecl *RD) {
```

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaDeclAttr.cpp:9330]: What do you mean here, exactly?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (gmlueck) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:11528]: Kind of a nit, but we are trying to encourage people to use the `sycl::` namespace rather than `cl::sycl::`.  Can we change the wording to say `... of type sycl::aspect` instead?

File: clang/test/CodeGenSYCL/device_has.cpp
Comment (gmlueck) [inline clang/test/CodeGenSYCL/device_has.cpp:47]: Should there be some tests in here for the case when the parameter to `sycl::device_has()` is a `constexpr` expression, rather than a literal value?  For example, something like:

```
constexpr cl::sycl::aspect get_aspect() {return cl::sycl::aspect::fp16;}
[[sycl::device_has(get_aspect())]] void func6();
```

File: clang/include/clang/Basic/AttrDocs.td
Comment (elizabethandrews) [reply clang/include/clang/Basic/AttrDocs.td:450]: Oh nice catch. Will change that. Thanks!

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaDeclAttr.cpp:9330]: I was wondering if there was a better way to do this (like `LookUpResult ` and compare or something) but it doesn't make sense to do that here. So I will be keeping this code and removing this comment.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaDeclAttr.cpp:9344]: Yep. I took the logic from there. I haven't looked too closely yet, but to avoid duplication I think we will need to move the `Util `class in `SemaSYCL `to `Sema `or something. Is it alright with you if I do that in a follow-patch?  I think it will add a lot of unrelated refactoring changes in `SemaSYCL `to this PR

File: clang/include/clang/Basic/AttrDocs.td
Comment (premanandrao) [inline clang/include/clang/Basic/AttrDocs.td:443]: Is the second part implemented?

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (premanandrao) [inline clang/lib/Sema/SemaDeclAttr.cpp:9331]:
```suggestion
// code duplication.
```

**Review by premanandrao — COMMENTED** [review]
Mostly looks good to me.   Just a couple nits.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [inline clang/lib/Sema/SemaDeclAttr.cpp:9373]: I have one question here: Are we ignoring duplicate attributes here? Could you please add test to make sure that we ignore duplicate attribute here.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [inline clang/lib/Sema/SemaDeclAttr.cpp:9406]: Same question here for duplicate attributes?

File: clang/include/clang/Basic/AttrDocs.td
Comment (elizabethandrews) [reply clang/include/clang/Basic/AttrDocs.td:443]: No. These diagnostics will be emitted by a post-link tool as specified here - https://github.com/intel/llvm/blob/sycl/sycl/doc/OptionalDeviceFeatures.md

File: clang/test/CodeGenSYCL/device_has.cpp
Comment (elizabethandrews) [reply clang/test/CodeGenSYCL/device_has.cpp:28]: func6 is a redeclaration. I added the test to ensure attributes are inherited on redeclarations

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaDeclAttr.cpp:9373]: Yes duplicate attributes are ignored. I added a test to show warning but I can also add an AST test if you like

Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:9373]: Could you please add AST test? Thanks.


https://github.com/intel/llvm/blob/975e348e4b2fe2f579f97171b252d3930a6b55e8/clang/lib/Sema/SemaDeclAttr.cpp#L6306-#L6307

Comment (elizabethandrews) [reply clang/lib/Sema/SemaDeclAttr.cpp:9373]: Added an AST test. When writing test I realized I hadn't handled duplicate attribute on same declaration and so added code for that and added a CodeGen test

**Review by smanna12 — DISMISSED** [review]
Changes look good to me.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (Fznamznon) [reply clang/lib/Sema/SemaDeclAttr.cpp:9345]: >  Is it alright with you if I do that in a follow-patch?
Sure!

Comment (elizabethandrews) [reply clang/lib/Sema/SemaDeclAttr.cpp:9345]: https://github.com/intel/llvm/issues/5186

File: clang/lib/Basic/Attributes.cpp
Comment (Fznamznon) [inline clang/lib/Basic/Attributes.cpp:14]: Just for my education, did I understand it correctly that both forms `[[sycl_detail::uses_aspects()]]` and `[[__sycl_detail__::__uses_aspects__()]]` are accepted?

**Review by gmlueck — APPROVED** [review]

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:675]: These function are quite similar to each other and to `instantiateIntelFPGABankBitsAttr` (and maybe more) described below. That is just a general observation, no need to do anything in this concrete patch.

File: clang/test/CodeGenSYCL/device_has.cpp
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/device_has.cpp:28]: I assume it is redeclaration of `func6` declared first time on line 26.

**Review by Fznamznon — APPROVED** [review]
No more concerns from my side.

File: clang/test/CodeGenSYCL/device_has.cpp
Comment (premanandrao) [reply clang/test/CodeGenSYCL/device_has.cpp:28]: Isn't that the case with func5 above too?

Comment (Fznamznon) [reply clang/test/CodeGenSYCL/device_has.cpp:28]: Fair enough :) seems to be the same test case

Comment (elizabethandrews) [reply clang/test/CodeGenSYCL/device_has.cpp:28]: I didn't notice `func5 `was a redeclaration. I'll change that. Thanks for catching it!

File: clang/lib/Basic/Attributes.cpp
Comment (elizabethandrews) [reply clang/lib/Basic/Attributes.cpp:14]: Right now no.  Prior to making this change, I expected both to work. But for some reason __uses_aspects__ just did not work. I tried debugging it for sometime to see why it was not getting normalized correctly but could not figure it out. I then noticed that this file has many other functions doing normalization and it is only done for gnu and clang namespace. So I added a guard in this function to not normalize for `__sycl_detail__ ` namespace.

Comment (elizabethandrews) [reply clang/lib/Basic/Attributes.cpp:17]: I added a FIXME here because I am curious about why this is not working. Since the specification only requires `__uses_aspects__` to work I think adding this guard for now is fine, but for upstreaming we may need to figure this out at some point.

**Review by smanna12 — APPROVED** [review]

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:675]: Yes we used to have a common templated function for many of these but we changed it to be separate functions to be in line with how community handles attributes.

File: clang/lib/Basic/Attributes.cpp
Comment (Fznamznon) [reply clang/lib/Basic/Attributes.cpp:17]: Got that, thanks. I'm not sure we even need "uses_aspects" to be supported, since the design doc (github.com/intel/llvm/blob/sycl/sycl/doc/OptionalDeviceFeatures.md) says that it should be spelled with double underscores in DPC++ headers anyway. Just asked for better understanding.
I would assume then that to make both forms work we should declare the attribute in `Attr.td` without underscores and then its name will be normalized somewhere in this file.

Comment (elizabethandrews) [reply clang/lib/Basic/Attributes.cpp:17]: I tried that but it didn't work. Not sure if I made some unintentional mistake though

**Review by Fznamznon — APPROVED** [review]

**Comment by elizabethandrews** [comment]
Thanks for the reviews everyone!

**Review by premanandrao — APPROVED** [review]

**Comment by elizabethandrews** [comment]
@bader is anything holding this up? I believe I've received the required approvals and tests pass.

**Pull request merged**
